storage/src/freezer/storage.rs

```diff
@@ -0,0 +1,624 @@
+impl<K: Array + Codec<Cfg = ()>, V: Codec> commonware_codec::Write for StorageValue<K, V> {
+    fn write(&self, buf: &mut impl bytes::BufMut) {
+        match self {
+            StorageValue::Value(v) => {
+                buf.put_u8(0);
+                v.write(buf);
+            }
+            StorageValue::Key(k) => {
+                buf.put_u8(1);
+                k.write(buf);
+            }
+        }
+    }
+}
+
+impl<K: Array + Codec<Cfg = ()>, V: Codec> commonware_codec::Read for StorageValue<K, V> {
+    type Cfg = V::Cfg;
+
+    fn read_cfg(
+        buf: &mut impl bytes::Buf,
+        cfg: &Self::Cfg,
+    ) -> Result<Self, commonware_codec::Error> {
+        let tag = buf.get_u8();
+        match tag {
+            0 => Ok(StorageValue::Value(V::read_cfg(buf, cfg)?)),
+            1 => Ok(StorageValue::Key(K::read_cfg(buf, &())?)),
+            _ => Err(commonware_codec::Error::InvalidEnum(tag)),
+        }
+    }
+}
+
+/// Implementation of `Freezer` storage using a single diskmap with prefixed keys.
+pub struct Freezer<E: Storage + Metrics, K: Array, V: Codec> {
+    // Single DiskMap for both key->value and index->key storage
+    // We use a large enough array to hold any prefixed key (1 byte prefix + max of key or index)
+    storage: DiskMap<E, PrefixedArray<256>, StorageValue<K, V>>,
+
+    // RMap for interval tracking (only in-memory data structure)
+    intervals: RMap,
+
+    // Metrics
+    items_tracked: Gauge,
+    gets: Counter,
+    has: Counter,
+    syncs: Counter,
+
+    // Store the actual key size for runtime validation
+    _key_size: usize,
+}
+
+impl<E: Storage + Metrics, K: Array + Codec<Cfg = ()>, V: Codec> Freezer<E, K, V> {
+    /// Initialize a new `Freezer` instance.
+    ///
+    /// The in-memory RMap is populated during this call by scanning the index entries.
+    pub async fn init(context: E, cfg: Config<V::Cfg>) -> Result<Self, Error> {
+        // Validate that our key size fits in the prefixed array
+        let key_size = K::SIZE;
+        if key_size + 1 > 256 {
+            return Err(Error::Runtime(commonware_runtime::Error::PartitionCorrupt(
+                format!("Key size {} is too large (max 255)", key_size),
+            )));
+        }
+
+        // Initialize single diskmap for all storage
+        let diskmap_config = DiskMapConfig {
+            partition: cfg.partition.clone(),
+            directory_size: cfg.directory_size,
+            codec_config: cfg.codec_config.clone(),
+            write_buffer: cfg.write_buffer,
+            target_journal_size: cfg.target_journal_size,
+        };
+        let storage: DiskMap<E, PrefixedArray<256>, StorageValue<K, V>> =
+            DiskMap::init(context.with_label("storage"), diskmap_config).await?;
+
+        // Initialize RMap and rebuild from existing data
+        let mut intervals = RMap::new();
+        {
+            debug!("initializing freezer");
+
+            // Replay the diskmap to rebuild the RMap from index entries
```

Comment:
note that this is required to ensure safety during unclean recovery (want to ensure keys can be accessed by index and key).

May be a way to avoid full replay of all data...